### PR TITLE
feat: inc parts of semver, change snapshot default name

### PR DESCRIPTION
**Pull request by caarlos0**
revives https://github.com/goreleaser/goreleaser/pull/1457 adding some new template functions:

- incpatch
- incminor
- incmajor

and change the `goreleaser init`  generated config snapshot name to `{{ incpatch .Tag }}-next`

refs https://github.com/goreleaser/goreleaser/issues/2355

### CODE REVIEW

File: internal/pipe/git/git.go [hunk 1]
```diff
@@ -32,7 +32,9 @@ func (Pipe) Run(ctx *context.Context) error {
 		return err
 	}
 	ctx.Git = info
-	log.Infof("releasing %s, commit %s", info.CurrentTag, info.Commit)
+	if !ctx.Snapshot {
+		log.Infof("releasing %s, commit %s", info.CurrentTag, info.Commit)
```
Comment (radeksimko) [inline internal/pipe/git/git.go:36]: I agree that _releasing_ is not relevant here, but wouldn't it be still useful to output the commit SHA as a confirmation of what exact SHA did GoReleaser pick up?

**Review by radeksimko — COMMENTED** [review]
I relate to @myitcv in #2355 in the sense that the current name with `major.minor.patch-SNAPSHOT-commit` is not very conventional but I still believe that there's value in it being different from a regular "stable" release. This PR would make the two indistinguishable which I would be unfortunate IMO.

How about changing it something like `major.minor.patch-prerelease.commit` or ``major.minor.patch-prerelease` or anything similar that follows semver more closely?

https://semver.org/#spec-item-9

> A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version.

**Comment by radeksimko** [comment]
Apologies for bit of a trigger happiness with the last comment, after re-reading the discussion in https://github.com/goreleaser/goreleaser/issues/2355 I suppose that the problem you're all trying to solve is predictability of the name of the resulting artifact.

I obviously can't speak for any majority of users, but from my experience this has not been a common concern for us at HashiCorp. I understand the CI-testability use case, but a real concern I would have if snapshots were named the same as the upcoming version is that someone will accidentally publish that as a "stable" version and then override with a real stable release later.

Having the commit SHA there is from my experience valuable to make releasing more error-prone and to easily set expectations for end-users _by default_.

I suppose that one group of users will have to end up specifying this manually. I would personally be more inclined to keep the _default_ options _safer_.

https://goreleaser.com/customization/snapshots/

**Comment by myitcv** [comment]
@radeksimko I'm very inexperienced with `goreleaser`, so my comments should be taken with a pinch of salt!

> but a real concern I would have if snapshots were named the same as the upcoming version is that someone will accidentally publish that as a "stable" version and then override with a real stable release later.

Please can you explain this point a bit further? Because it's entirely possible I'm not entirely following, or I have a different mental model. I expect `goreleaser test` to, by default, leave no trace of what it did locally. No files, no docker images, nothing. It goes without saying this mode would also not upload/push anything, ever. For those people who do want to test a docker image, perhaps a flag `--nocleanup` could be provided... but at this point I'm getting out of my depth!

My desire to have `goreleaser test` accept a version string is that it allows us to simulate the entire release process without actually mutating any remote state (GitHub release assets, docker images etc). The stable output of `goreleaser test` would allow us to affirm what _would_ happen if we were to run things for real. Because without that, the only time we can truly test what will be published (in terms of asset names) is when it comes to cutting a release. That just feels fragile, especially if you consider `goreleaser` updates along the way (which will hopefully be stable!)

But again, I really defer to those with greater experience of `goreleaser`. Please consider my experience reports as just that: I'm very happy being guided to doing things a different way, just so long as it gives us comfort with our `goreleaser` setup.

**Comment by radeksimko** [comment]
> I expect goreleaser test to, by default, leave no trace of what it did locally. No files, no docker images, nothing. It goes without saying this mode would also not upload/push anything, ever.

That more or less aligns with my thinking but my this PR would change behaviour of the `goreleaser build --snapshot` and `goreleaser release --snapshot` commands. I would be more than happy for a `test` command to behave as you described.

I do believe the `test` command should perhaps use different (default?) naming convention for the snapshot as it has different use cases.

**Comment by caarlos0** [comment]
Yeah, I think that changing the default here is indeed no good.

Maybe we could instead have a different field for the actual `test` command later, which will actually be something @myitcv is interested in as well, e.g. `test.tag_template: v99.9.9` (which by default could be `{{ inpatch .Tag }}`).

wdyt?

**Comment by caarlos0** [comment]
should be better now.

I did change the default on the config generated by `goreleaser init`, which shiould not a breaking change, but will be a bit better than the previous value `TAG-next`...

**Review by radeksimko — APPROVED** [review]
LGTM 🚢

**Pull request merged**